src/libcore/slice/memchr.rs

```diff
@@ -102,8 +102,24 @@ pub fn memrchr(x: u8, text: &[u8]) -> Option<usize> {
     let ptr = text.as_ptr();
     let usize_bytes = mem::size_of::<usize>();
 
+    // a version of align_offset that says how much must be *subtracted*
+    // from a pointer to be aligned.
+    #[inline(always)]
+    fn align_offset_down(ptr: *const u8, align: usize) -> usize {
+        let align_offset = ptr.align_offset(align);
+        if align_offset > align {
+            // Not possible to align
+            usize::max_value()
+        } else if align_offset == 0 {
+            0
+        } else {
+            // E.g. if align=8 and we have to add 1, then we can also subtract 7.
+            align - align_offset
```

Comment:
You probably want `(align - align_offset) & (align - 1)` (or `% align` if you're sure it will optimize) instead of the last if/else. Not sure about the first one though, seems kinda fine as-is.